api/testing/restapi_test.go

```diff
@@ -0,0 +1,686 @@
+package testing
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/libopenstorage/openstorage/api"
+	client "github.com/libopenstorage/openstorage/api/client"
+	volumeclient "github.com/libopenstorage/openstorage/api/client/volume"
+	"github.com/libopenstorage/openstorage/api/server"
+	"github.com/libopenstorage/openstorage/volume"
+	"github.com/stretchr/testify/assert"
+	"go.pedge.io/dlog"
+
+	"github.com/golang/mock/gomock"
+	"github.com/kubernetes-csi/csi-test/utils"
+
+	mockcluster "github.com/libopenstorage/openstorage/cluster/mock"
+	volumedrivers "github.com/libopenstorage/openstorage/volume/drivers"
+	mockdriver "github.com/libopenstorage/openstorage/volume/drivers/mock"
+)
+
+// testServer is a simple struct used abstract
+// the creation and setup of mock server
+type testServer struct {
+	client *client.Client
+	m      *mockdriver.MockVolumeDriver
+	c      *mockcluster.MockCluster
+	mc     *gomock.Controller
+}
+
+const (
+	host       string = "http://127.0.0.1:2376"
+	mgmtPort   uint16 = 2376
+	pluginPort uint16 = 2377
+	driver     string = "mock"
+	version    string = "v1"
+)
+
+// Init function to setup the http server
+
+func init() {
+	startServer()
+}
+
+func startServer() {
+	err := server.StartVolumeMgmtAPI(
+		driver,
+		volume.DriverAPIBase,
+		mgmtPort,
+	)
+
+	if err != nil {
+		dlog.Errorf("Error starting the server")
+	}
+
+	// adding sleep to avoid race condition of connection refused.
+	time.Sleep(1 * time.Second)
+}
+
+func setupMocks() *testServer {
+
+	var ts = &testServer{}
+
+	// Add driver to registry
+	ts.mc = gomock.NewController(&utils.SafeGoroutineTester{})
+	ts.m = mockdriver.NewMockVolumeDriver(ts.mc)
+	ts.c = mockcluster.NewMockCluster(ts.mc)
+
+	err := volumedrivers.Add(driver, func(map[string]string) (volume.VolumeDriver, error) {
+		return ts.m, nil
+	})
+
+	if err != nil {
+		dlog.Errorf("Failed to add the driver [%s] for tests", driver)
+	}
+
+	// Register the mock driver
+	err = volumedrivers.Register(driver, nil)
+
+	return ts
+}
+
+func TestServerStart(t *testing.T) {
+
+	ts := setupMocks()
+	defer ts.Stop()
+
+	var err error
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	assert.Nil(t, err)
+	assert.NotNil(t, ts.client)
+}
+
+func TestVolumeCreateSuccess(t *testing.T) {
+
+	var err error
+	ts := setupMocks()
+	defer ts.Stop()
+
+	// create a request
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	assert.Nil(t, err)
+	assert.NotNil(t, ts.client)
+
+	// Setup request
+	name := "myvol"
+	size := uint64(1234)
+
+	req := &api.VolumeCreateRequest{
+		Locator: &api.VolumeLocator{Name: name},
+		Source:  &api.Source{},
+		Spec:    &api.VolumeSpec{Size: size},
+	}
+
+	// Setup mock functions
+	id := "myid"
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Create(gomock.Any(), gomock.Any(), gomock.Any()).
```

Comment:
Please check to see which values can be matched here if any. It provides more testing than allowing _any_ value. Gomock is very good at matching structs.